gef-ui/server/controllers/facility-end-date/index.ts

```diff
@@ -28,61 +59,71 @@ export const getFacilityEndDate = async (req: GetFacilityEndDateRequest, res: Re
-      dealId,
-      facilityId,
-      status,
-    };
-
-    if ('facilityEndDate' in facility && typeof facility.facilityEndDate === 'string') {
-      const facilityEndDate = parseISO(facility.facilityEndDate);
-      facilityEndDateViewModel.facilityEndDate = {
-        day: String(getDate(facilityEndDate)),
-        month: String(getMonth(facilityEndDate) + 1),
-        year: String(getYear(facilityEndDate)),
-      };
+    if (shouldRedirectFromPage) {
+      return res.redirect(previousPage);
     }
 
-    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+    return res.render('partials/facility-end-date.njk', getFacilityEndDateViewModel(facility, previousPage, status));
   } catch (error) {
     console.error(error);
     return res.render('partials/problem-with-service.njk');
   }
 };
 
-const getCoverStartDateOrStartOfToday = (facility: Record<string, unknown>): Date => {
-  if (typeof facility.coverStartDate === 'string') {
-    return startOfDay(parseISO(facility.coverStartDate));
-  }
+export const getFacilityEndDateFromUnissuedFacilitiesPage = async (req: GetFacilityEndDateRequest, res: Response) =>
+  handleGetFacilityEndDate(req, res, `/gef/application-details/${req.params.dealId}/unissued-facilities/${req.params.facilityId}/about`);
 
-  if (!facility.coverStartDate) {
-    return startOfDay(new Date());
-  }
+export const getFacilityEndDateFromApplicationPreviewPage = async (req: GetFacilityEndDateRequest, res: Response) =>
+  handleGetFacilityEndDate(req, res, `/gef/application-details/${req.params.dealId}/unissued-facilities/${req.params.facilityId}/change`);
+
+export const getFacilityEndDateFromApplicationDetailsPage = async (req: GetFacilityEndDateRequest, res: Response) =>
+  handleGetFacilityEndDate(req, res, `/gef/application-details/${req.params.dealId}/facilities/${req.params.facilityId}/about-facility`);
 
-  throw new Error('Invalid coverStartDate');
+const updateFacilityEndDateIfChanged = async (
+  existingFacility: Record<string, unknown>,
+  facilityEndDate: Date,
+  userSession: LoggedInUserSession,
+): Promise<void> => {
+  const facilityEndDateNeedsUpdating =
+    typeof existingFacility.facilityEndDate !== 'string' || !isSameDay(parseISO(existingFacility.facilityEndDate), facilityEndDate);
+
+  if (!facilityEndDateNeedsUpdating) {
+    return;
+  }
+  await api.updateFacility({
+    facilityId: existingFacility._id,
+    payload: {
+      facilityEndDate,
+    },
+    userToken: userSession.userToken,
+  });
+
+  const applicationUpdate = {
+    editorId: userSession.user._id,
+  };
+  await api.updateApplication({ dealId: existingFacility.dealId, application: applicationUpdate, userToken: userSession.userToken });
 };
 
-export const postFacilityEndDate = async (req: PostFacilityEndDateRequest, res: Response) => {
+const handlePostFacilityEndDate = async (req: PostFacilityEndDateRequest, res: Response, uris: PostRequestUris) => {
```

Comment:
maybe we could split this in future to make this file less large

Reply:
remind me in 7219 and I will